### PR TITLE
Support arrays in BindIn, besides Collection #473

### PR DESCRIPTION
**Pull request by TheRealMarnes**
https://github.com/jdbi/jdbi/issues/473

### CODE REVIEW

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 1]
```diff
@@ -85,6 +86,42 @@ public void apply(SQLStatement q) throws SQLException
                 }
             };
         }
+
+		private static Collection<?> tryToCast(Object arg) {
+			if (arg instanceof Collection) {
+				return (Collection<?>) arg;
+			}
+			if (arg instanceof Object[]) {
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: I'd probably like to see this the other way around: convert collections to an array and bind from the array. You can use `java.lang.reflect.Array.getLength(array)` and `Array.get(array, index)` and then you don't have to have all this code testing for different array types.

Also, there's a checkstyle error failing the build, because your code editor is using tabs instead of spaces for indentation. This project uses spaces.

**Comment by TheRealMarnes** [comment]
@qualidafial sorry, I only meant the pull request as a rough draft that needs to be fine-tuned before fully merging, not as a ready-to-go submission... There's a million aspects to how a piece of code is written, like logging, exception catching and throwing, message style, considerations like which situations to check first, or like you said indeed maybe to use an array instead of a list entirely, etc. I can't possibly know how you guys like the code to be at every level, just wanted to write this as a quick POC

**Comment by TheRealMarnes** [comment]
Isn't using reflection kinda bad btw? It's slow and hacky, while the type check may be verbose but it's fast and clean

**Comment by qualidafial** [comment]
Java reflection hasn't been "slow" for years. Certainly it's slower than static code, but not by the crippling factor that it used to be. Moreover as seldom as `@BindIn` is used in practice, this case is not likely to become a performance bottleneck.

If you can demonstrate that use of reflection here adversely affects performance in real-world scenarios, I'd be happy to revisit. But absent any evidence of performance issues, I'd rather use code that is maintainable and easy to understand.

**Comment by TheRealMarnes** [comment]
Fair enough, I'll patch it up some more. I'm not saying it'll be perfect though, someone who's more in on things in jdbi's specific domain should finetune it thoroughly anyway

**Comment by qualidafial** [comment]
Just a heads up: unless a project is using compile-time annotation processors (e.g. Dagger DI), most libraries that make use of annotations are doing so through reflection. Nearly all annotation-driven APIs I'm aware of do their magic through reflection--JDBI included.

**Comment by stevenschlansker** [comment]
That's a little different though.  Accessing annotations via reflection is once per statement, whereas mapping like this will be once per value converted.  So it will be many orders of magnitude more common.  I still lean towards the readable version first, then we can iterate and improve it as we find bottlenecks, but it's quite reasonable to consider the performance cost.  And thanks @TheRealMarnes for giving this a go, don't worry at all if you are iteratively improving it -- that's what code review is for :)

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 2]
```diff
@@ -85,6 +86,42 @@ public void apply(SQLStatement q) throws SQLException
...
+				return Arrays.asList((Object[]) arg);
+			}
+			// necessary because of how java handles primitives...
+			if (arg instanceof boolean[]) {
+				return Arrays.asList((boolean[]) arg);
+			}
+			if (arg instanceof byte[]) {
+				return Arrays.asList((byte[]) arg);
```
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: Are you sure this construct actually works?  I believe this is actually creating a `List<byte[]>` which is not what you want.

**Comment by stevenschlansker** [comment]
Needs some test cases.  I suspect that adding tests will reveal that the `Arrays.asList` approach won't work -- it won't do the dirty work of boxing primitives or the like.

**Comment by TheRealMarnes** [comment]
I'm going to put the casting/reflection utility method in a utility class so it can be cleanly tested. Where do you guys want it to be (package and/or class)?

**Comment by stevenschlansker** [comment]
I think there's a `.util` package that might naturally host it.  Other than that, no particular preference, just come up with a good name for it :)

**Comment by TheRealMarnes** [comment]
Dunno, that package seems to be (and the package.html even says so) meant for utilities for the end users. This utility is something that should stay out of the end user's way. Maybe I should make a util_private package?

**Comment by stevenschlansker** [comment]
Or just a different class in the same package -- that way you can make it package-private and not visible to end users.

**Comment by TheRealMarnes** [comment]
Not the cleanest from a project files pov, but that does make it completely invisible to users. I like that idea more :P

**Comment by TheRealMarnes** [comment]
For some reason I can't get the code to compile locally, in intellij or maven, because of missing dependencies (org.skife.jdbi.rewriter package among others). I'll just push and see if jenkins starts building it or anything. I'm not very familiar with github or this project

File: src/main/java/org/skife/jdbi/v2/unstable/Objects.java [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    private Objects() {}
+
+    static Object[] toArray(final Object obj) {
+        if (obj instanceof Collection) {
+            return ((Collection) obj).toArray();
+        }
+
+        if (obj.getClass().isArray()) {
+            final int length = Array.getLength(obj);
+            final Object[] out = new Object[length];
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/unstable/Objects.java]: You only need to copy the array if `obj.getClass().getComponentType().isPrimitive()`. Otherwise you can just cast to `Object[]`

**Comment by TheRealMarnes** [comment]
Thanks for the catch, I'd forgotten to re-add that

**Comment by TheRealMarnes** [comment]
Come to think of it, why not support Iterable<?>?

**Comment by stevenschlansker** [comment]
Supporting `Iterable` makes a lot of sense.  You could even consider supporting `Iterator` too, although this starts to get more complex :)

Your build errors due to missing `rewriter` package is likely because IntelliJ doesn't know to build the ANTLR grammars -- this is handled in the `generate-sources` Maven phase: https://github.com/jdbi/jdbi/blob/master/pom.xml#L321-L332  So I'm a little surprised that Maven doesn't work -- what command do you invoke?  what error do you get?

Eclipse knows how to figure this out (with a plugin) but unfortunately I don't know how to get IntelliJ set up right.

**Comment by TheRealMarnes** [comment]
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-toolchains-plugin:1.1:toolchain (default) on project jdbi: Cannot find matching toolchain definitions for the following toolchain types:
[ERROR] jdk [ vendor='sun' version='1.6' ]
[ERROR] Please make sure you define the required toolchains in your ~/.m2/toolchains.xml file.

Yay... but wait, maven says java 6, the travis file says jdk7? Maybe the jdk version numbering is a little wonky and it's not like I've looked it up, but shouldn't it be the same?

**Comment by stevenschlansker** [comment]
Ah, yeah this needs to be documented better.  You have to configure the Maven Toolchains on your dev box:
http://maven.apache.org/plugins/maven-toolchains-plugin/usage.html
http://maven.apache.org/plugins/maven-toolchains-plugin/toolchains/jdk.html

Travis only has openjdk7 type, but it actually has Java 6 still -- and jdbi v2 is based on Java 6, so we use Toolchains to make sure you use the right one:
https://github.com/jdbi/jdbi/blob/master/src/build/travis-toolchains.xml

**Comment by arteam** [comment]
Yes, that is a rather confusing message. You need to set up Maven
toolchains locally to point to a JDK which can compile Java 6 code. I think
we moved this functionality to a specific profile in JDBI3, just for the
reason a to simplify the development setup for the people who want to
contribute.
Am 06.09.2016 22:01 schrieb "TheRealMarnes" <notifications@github.com>:

**Comment by stevenschlansker** [comment]
Also worth noting -- if you don't have Java 6 installed, you don't actually have to.  You can point your jdk6 toolchain at a Java 8 just fine.  The downside is if you accidentally use any new features, you won't notice on a local build - but Travis will catch any errors you make.

**Comment by TheRealMarnes** [comment]
Yeah, pointing it at my jdk8 is what I'm gonna go with...

**Comment by TheRealMarnes** [comment]
Ok, now I run into runtime type exceptions when I run a full test of @BindIn (with varargs for starters)... That'll be for another time :/

**Comment by stevenschlansker** [comment]
Please leave a note of any combinations you find that don't work!  We may not fix them all up for jdbi v2, but we are reworking a lot of this for v3, so we'd appreciate failing tests as issues or even a PR for the v3 branch 😄

**Comment by TheRealMarnes** [comment]
For some reason I can't get a screenshot attached here... Anyway, I've pushed the test class for BindIn, it passes on Iterable but fails on varargs and Iterator. The error is

java.lang.ClassCastException: [I cannot be cast to java.lang.Iterable
	at org.skife.jdbi.v2.unstable.BindIn$BindingFactory$1.bind(BindIn.java:98)

**Comment by TheRealMarnes** [comment]
For some reason I can't get a screenshot attached here... Anyway, I've pushed the test class for BindIn, it passes on Iterable but fails on varargs and Iterator. The error is

java.lang.ClassCastException: [I cannot be cast to java.lang.Iterable
	at org.skife.jdbi.v2.unstable.BindIn$BindingFactory$1.bind(BindIn.java:98)

**Comment by TheRealMarnes** [comment]
For some reason I can't get a screenshot attached here... Anyway, I've pushed the test class for BindIn, it passes on Iterable but fails on varargs and Iterator. The error is

java.lang.ClassCastException: [I cannot be cast to java.lang.Iterable
	at org.skife.jdbi.v2.unstable.BindIn$BindingFactory$1.bind(BindIn.java:98)

**Comment by qualidafial** [comment]
Something's broken in Github's UI so I can't leave a comment on the specific line, but in BindIn.java:

* Line 64: the contents of `argArray` are never used--just the length. The conversion to array is wasted here--it would be better to have an `Objects.size(arrayOrIterable)` method
* Lines 97-107: This is where the rubber hits the road for `@BindIn`: this is where you need to iterate  the iterable / array and bind each value to the named parameter introduced in `BindIn.CustomizerFactory.createForParameter`. Currently the code is still assuming the object is an `Iterable`, but not checking for arrays.

**Comment by qualidafial** [comment]
Something's broken in Github's UI so I can't leave a comment on the specific line, but in BindIn.java:

* Line 64: the contents of `argArray` are never used--just the length. The conversion to array is wasted here--it would be better to have an `Objects.size(arrayOrIterable)` method
* Lines 97-107: This is where the rubber hits the road for `@BindIn`: this is where you need to iterate  the iterable / array and bind each value to the named parameter introduced in `BindIn.CustomizerFactory.createForParameter`. Currently the code is still assuming the object is an `Iterable`, but not checking for arrays.

**Comment by TheRealMarnes** [comment]
Isn't a cast just a manual override of the reference type that doesn't actually involve any real work (like a true conversion would)?

**Comment by qualidafial** [comment]
If you're talking about array casts, you're mostly correct. Array casts are checked at runtime, e.g. you can cast a `String[]` to `Object[]`, and an `Object[]` reference that points to a `String[]` can be converted back to `String[]`. However casting a `int[]` to `long[]` will throw a `ClassCastException`.

**Comment by stevenschlansker** [comment]
Notably, primitive arrays do not descend from `Object[]` -- there is no common type to `String[]` and `int[]` other than `Object` itself.  That's why the `java.lang.reflect.Array` methods all take `Object` rather than an array type, and is also one reason why writing code to handle both is otherwise quite painful.  Java Arrays are covered in warts...

**Comment by TheRealMarnes** [comment]
Yes, I know,  it's just... sorry, I'm having a big brainfart atm (6 hours of sleep in total over 2 nights might be related to that). I lost track since Matthew's comment about the line numbers, but I just dug around again, looked a bit further than my nose (I'm notorious for not doing that) and checked things out in debug mode again, and I think I my train of thought is back on rails.

What's happening is that Bindifier.bind is trying to pass BindIn's arguments (varargs, iterator, ...) to Binder.bind, which expects an iterable, and that's where it goes wrong. I think I was being misled a bit by the stacktrace, I'm used to writing code in such a way that type errors are impossible even at runtime, so it didn't occur to me the cast exception was coming from the method parameters themselves, and that's how I lost track.

**Comment by TheRealMarnes** [comment]
It works for varargs :) Now to make it work with iterator

**Comment by TheRealMarnes** [comment]
Aaahhh and now it's failing on iterator with "no parameter found for placeholder", because the iterator arg becomes an empty array... BindIn first checks its size and later has to iterate it again to get the contents, but checking its size already spent its iteration...

**Comment by stevenschlansker** [comment]
Yeah, maybe not worth it.  Feel free to cut `Iterator` support, I think `Iterable` is enough.  If you actually want to do `Iterator` you probably have to dump it into a temporary array, which is not great.

**Comment by TheRealMarnes** [comment]
Yeah, since iterators can't be reset and we can't really link the 2 parts of BindIn to share the original params, I think there's logically no way around it. I don't think there's much of a point to it anyway since the iterator needs to be completely iterated to fill up the query, so it's not like using an iterator instead of a collection/array is saving anything.

File: src/main/java/org/skife/jdbi/v2/unstable/Objects.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+                return out;
+            }
+        }
+
+        throw new IllegalArgumentException("argument must be one of the following: Iterable, or an array/varargs (primitive or complex type)");
+    }
+
+    static int arraySize(final Object obj) {
+        if (obj instanceof Iterable) {
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/unstable/Objects.java]: It may be worthwhile to check for `Collection` first, and return `((Collection)obj).size()`.

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 5]
```diff
@@ -25,82 +25,72 @@
...
+                ids.add("__" + key + "_" + i);
             }
 
-            StringBuilder names = new StringBuilder();
-            for (Iterator<String> i = ids.iterator(); i.hasNext();) {
-                names.append(":").append(i.next());
-                if (i.hasNext()) {
+            final StringBuilder names = new StringBuilder();
+            for (final Iterator<String> it = ids.iterator(); it.hasNext(); ) {
+                names.append(":").append(it.next());
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: This is sort of orthogonal to this PR, but it would create less objects if the previous for-loop were collapsed into this one:

    for (int i = 0; i < size; i++) {
      if (id > 0) {
        names.append(",");
      }
      names.append(":__").append(key).append('_').append(i);
    }

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 6]
```diff
@@ -25,82 +25,72 @@
...
-        public SqlStatementCustomizer createForMethod(Annotation annotation, Class sqlObjectType, Method method)
-        {
+        public SqlStatementCustomizer createForMethod(final Annotation annotation, final Class sqlObjectType, final Method method) {
             throw new UnsupportedOperationException("Not supported on method!");
         }
 
         @Override
-        public SqlStatementCustomizer createForType(Annotation annotation, Class sqlObjectType)
-        {
+        public SqlStatementCustomizer createForType(final Annotation annotation, final Class sqlObjectType) {
```
Comment (qualidafial) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: While I share your love of K&R brace style, we're standardized on Allman brace style (open brace on the next line) on this project.

More to the point though, changing lines just for formatting makes the real history of changes opaque when you need to see the reason a line of code got to be like it is. Can you revert these lines that are just making whitespace changes?

File: src/test/java/org/skife/jdbi/v2/sqlobject/TestBindIn.java [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        in.add(2);
+
+        s.get(in.iterator());
+    }
+
+    @UseStringTemplate3StatementLocator
+    private interface SomethingIterator {
+        @SqlQuery("select id, name from something where id in (<ids>)")
+        List<Something> get(@BindIn("ids") Iterator<Integer> ids);
+    }
```
Comment (qualidafial) [inline src/test/java/org/skife/jdbi/v2/sqlobject/TestBindIn.java]: What happens if you pass an empty array/collection to a `@BindIn`?

**Comment by qualidafial** [comment]
Good tests.

**Comment by qualidafial** [comment]
👍 from me, after addressing my comment about whitespace changes, and adding a test to characterize what happens when you pass an empty array / iterable for a `@BindIn` parameter.

**Comment by TheRealMarnes** [comment]
Do we want to allow passing an empty argument to BindIn? Technically there's nothing wrong with it and neutral code that only crashes when it needs to is nice to have, but functionally it makes no sense and is probably an error the user wants to know about...

**Comment by TheRealMarnes** [comment]
I think things are done. Thoughts, anyone?

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 8]
```diff
@@ -90,21 +100,98 @@ public void apply(SQLStatement q) throws SQLException
...
+                    for (final Object s : arr)
+                    {
                         q.bind("__" + key + "_" + idx++, s);
                     }
                 }
             };
         }
     }
+
+    final class Util
```
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: nit: this class should be static

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 9]
```diff
@@ -90,21 +100,98 @@ public void apply(SQLStatement q) throws SQLException
...
+    {
+        private Util()
+        {
+        }
+
+        static Object[] toArray(final Object obj)
```
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: Could consider converting to `Iterator` instead here -- returning `Arrays.asList(obj).iterator()` for non-primitive arrays, `Iterable.iterator()` for iterators, and a lazy `Iterator` over `Array.get()` for primitive arrays.

That would save having to assemble the whole array at once, you would lazily iterate over it.

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 10]
```diff
@@ -90,21 +100,98 @@ public void apply(SQLStatement q) throws SQLException
...
+                    for (int i = 0; i < length; i++)
+                    {
+                        out[i] = Array.get(obj, i);
+                    }
+
+                    return out;
+                }
+            }
+
+            throw new IllegalArgumentException("argument must be one of the following: Iterable, or an array/varargs (primitive or complex type)");
```
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: Could consider appending `obj.getClass()` to the reason here, so you know what type was actually sent in, in case it isn't what you were expecting.

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 8 above)
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: isn't an inner class on an interface public static by default? Intellij points out that static is redundant when I try it, and classes outside the package can see BindIn.Util (suggesting the implicit public)

File: src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+import java.util.UUID;
+
+public class BindInTest
+{
+    private static Handle handle;
+
+    @BeforeClass
+    public static void init()
+    {
+        final JdbcDataSource ds = new JdbcDataSource();
```
Comment (stevenschlansker) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java]: Could consider the `H2DatabaseRule` instead of doing manual setup.

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 9 above)
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: What would that save though? We're going to iterate over the entire bunch of elements every time anyway

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 10 above)
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: good point

**Comment by stevenschlansker** [comment]
Nice tests!

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 8 above)
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: Oops, missed that it was on an annotation.  My bad.

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 9 above)
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: Consider binding a `List<MyBean>` -- right now, it'll assemble an intermediate `MyBean[]` which is only then used to iterate over to bind.  You could bind directly from the `Iterator` and remove the temporary copy.  It's a relatively small point, so it's an optional fixup if you like :)

**Comment by stevenschlansker** [comment]
Consider the comments above (mostly the one about building intermediate arrays), then LGTM!

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 9 above)
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: Hmm, yeah, I suppose it does create a little less cruft/shortcuts a few steps... I mildly dislike the idea of passing iterators around because I prefer method arguments to be "natural" and obvious rather than obfuscated/indirect, but if I'm honest with myself that's not much of a good argument anyway. I'll patch it

File: src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java (hunk 11 above)
Comment (TheRealMarnes) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java]: Sorry, what?

**Comment by TheRealMarnes** [comment]
Squash everything down to 1 commit then?

History gon' be rewritten guuuud...

**Comment by stevenschlansker** [comment]
I'm not a stickler for exactly one commit all the time, but having history polluted with multiple rounds of checkstyle fixups isn't really helpful :)

I'm a fan of "as many commits as reasonably make sense in isolation" -- which, for a relatively self-contained new feature like this, is usually one or two commits IMO.

File: src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java (hunk 11 above)
Comment (stevenschlansker) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java]: https://github.com/jdbi/jdbi/blob/jdbi3/core/src/test/java/org/jdbi/v3/core/H2DatabaseRule.java

used like this

https://github.com/jdbi/jdbi/blob/jdbi3/core/src/test/java/org/jdbi/v3/core/TestUuid.java#L26-L27

**Comment by TheRealMarnes** [comment]
Makes sense

*presses garbage compactor button*

**Comment by TheRealMarnes** [comment]
CustomizerFactory is final, but BindingFactory is not. Is this intentional?

File: src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java (hunk 11 above)
Comment (TheRealMarnes) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java]: but that's v3, this code is in v2

Comment (stevenschlansker) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java]: Gah, I'm sorry, I keep getting my wires crossed here...

Comment (TheRealMarnes) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInTest.java]: :P

**Comment by TheRealMarnes** [comment]
I added some null handling, exceptions and a tad of logging for the sake of not saddling users up with mysterious behavior. Is it overkill?

File: src/main/java/org/skife/jdbi/v2/unstable/ObjectIterator.java [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+            return Array.get(arr, index++);
+        } else {
+            throw new ArrayIndexOutOfBoundsException("only " + size + " elements available");
+        }
+    }
+
+    @Override
+    public void remove()
+    {
+        // no
```
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/ObjectIterator.java]: throw new UnsupportedOperationException()?

**Comment by qualidafial** [comment]
@stevenschlansker For future reference, we can do squash merges in Github without forcing contributors to squash changes by hand. It's an option under the dropdown on the `Merge pull request` button.

**Comment by qualidafial** [comment]
+1 from me

**Comment by arteam** [comment]
Great pull requests and review! +1 from me as well.

**Comment by TheRealMarnes** [comment]
It's not very important, but I just realized there's no test for Util.size(collection) and some of the test names are kinda poor

**Comment by TheRealMarnes** [comment]
Behavior of setting combinations of allowNull and allowEmpty needs to be tested and documented with a truth table

I'd suggest not merging yet, I'll work on this again tonight

**Comment by TheRealMarnes** [comment]
Is everyone alright with the functional ideas? I feel like I've made up half the functionality of BindIn on my own as I went along now, I'd like to make sure every piece of the design (not just technically but also functionally) is thought-out, deliberate, and agreed upon by the jdbi team.

The fact that there's a debug log statement for silent null passage (the only one in all of jdbi from what I can tell, slf4j will handle it safely with a noop logger in the worst case, but still) and an exception for unexpected behavior?

The separate options for allownull/empty: do we want the code to be strict or flexible automatically, or is configurable like this the best way? Should the default be strict instead of flexible? Not throwing exceptions unless really necessary is not a bad thing, but silent failures that get swept under the rug are just as bad. Where is the line between non-interrupting failure and throwing an IAE?

**Comment by qualidafial** [comment]
I feel like you went further with the `allowNull()` and `allowEmpty()` than I intended with my comment about characterization tests.

For background, there was a report (on the mailing lists I believe) a few weeks ago of a problem using `@BindIn` with an empty collection, which was apparently a valid use case for this specific user. I have not verified this, but the user reported that in Postgres you can't just say `WHERE foo IN ()`, it has to be `WHERE foo IN (null)` to represent an empty list.

So my intention was just to get some tests in place that capture whatever the current behavior is when the `@BindIn` parameter is empty. Sorry if this wasn't clear.

**Comment by TheRealMarnes** [comment]
Well, null and empty arguments need to be handled some way or other (at least definitely null), no? And since use cases vary, I basically wanted to make it possible to let the user choose between making them illegal cases or letting them pass quietly, with a sane default. If I removed that, we'd either have undefined behavior or unwanted exceptions.

Maybe what I could do to take it down a notch is combine the null and empty cases into a single case and provide a way to attach a nullness/emptiness handler the user can override to do what he wants and that to defaults to something sane? Thats what your report story inspires me to do anyway

My main concern was really just if with the growing complexity of BindIn I hadnt made any bad implementation/design choices, if maybe it wouldnt work make a better tool if it worked differently

**Comment by TheRealMarnes** [comment]
Tests added, now just thinking of ways to reduce the overengineering/cruft a bit

**Comment by TheRealMarnes** [comment]
So I got an idea yesterday to hopefully simplify things a bit, and came up with EmptyHandling (currently pushed and building, I can revert it if it's disliked)

Basically you'd use BindIn as
`@BindIn(value = "x", onEmpty = THROW)`
to throw an IAE, with VOID to just output empty parentheses "()" if the argument is null or empty, and NULL to output "(null)" (looking at you, postgres guy in the mailing list)

The code got kind of longer, I think, but in terms of complexity I think it got simpler

**Comment by TheRealMarnes** [comment]
I asked 2 friends who've been working as java devs for years to pretend to be end users and compare the booleans implementation to the new EmptyHandling one, and they both prefer the new one. I think we should stick to that. It seems cleaner, clearer and more extensible anyway, having an enum to express use cases.

Review/confirmation and thoughts, please? I didn't get to figuring out how to test the NULL case since I need to check the query itself for that, but I guess I'll find a way tonight. Maybe through a fake slf4j sql logger that asserts the query contents.

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java [hunk 13]
```diff
@@ -90,21 +127,120 @@ public void apply(SQLStatement q) throws SQLException
...
+
+        private static String getTypeWarning(final String type)
+        {
+            return "argument must be one of the following: Iterable, or an array/varargs (primitive or complex type); was " + type + " instead";
+        }
+    }
+
+    enum EmptyHandling
+    {
+        VOID, // output "()"
```
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: If these were Javadoc comments, it'd appear as an IDE tooltip.

File: src/main/java/org/skife/jdbi/v2/unstable/ObjectIterator.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+ */
+package org.skife.jdbi.v2.unstable;
+
+import java.lang.reflect.Array;
+import java.util.Iterator;
+
+/**
+ * Implements Iterator for unidentified arrays that have been cast to Object. Note that its elements will be returned as Object, primitives included (will be autoboxed).
+ */
+class ObjectIterator implements Iterator<Object>
```
Comment (stevenschlansker) [inline src/main/java/org/skife/jdbi/v2/unstable/ObjectIterator.java]: This class actually iterates over arrays, not any arbitrary `Object` (even though that's not proven at compile time) so IMO the name is a little confusing.  Maybe `ReflectionArrayIterator` or something like that?

File: src/main/java/org/skife/jdbi/v2/unstable/BindIn.java (hunk 13 above)
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/BindIn.java]: will do

File: src/main/java/org/skife/jdbi/v2/unstable/ObjectIterator.java (hunk 14 above)
Comment (TheRealMarnes) [inline src/main/java/org/skife/jdbi/v2/unstable/ObjectIterator.java]: agreed

**Comment by stevenschlansker** [comment]
This is looking great, the code is very readable!  Final comment: you could add a line to `RELEASE_NOTES` to announce the improvement.  Consider the last few comments, and then we can merge it.  Thanks for your persistence with this PR!

**Comment by TheRealMarnes** [comment]
Working on the final detail, but hit a snag. I'm writing two tests to verify that when setting onEmpty = NULL, the query does actually contain "null" where the placeholder was. I thought to use an SQLLog to do that, but I forgot the logger only gets a parameterized query with ? everywhere, without the parameter values. How can I get around this?

**Comment by qualidafial** [comment]
Since `@BindIn` is just defining what `<ids>` gets turned into, you could use a mock `Query` and verify that `Query.define("ids", expected)` gets called.

**Comment by TheRealMarnes** [comment]
Ah, right, I hadn't noticed a mock system was available, I'll take a look tonight, thanks

**Comment by TheRealMarnes** [comment]
The binding methods on SQLStatement are final... it's making mocking pretty difficult, but I might get there still.

**Comment by stevenschlansker** [comment]
If you need to make them not final, that's an OK change to make.

**Comment by TheRealMarnes** [comment]
Alright, thanks

**Comment by TheRealMarnes** [comment]
Well, I'm not quite happy with the EmptyHandling.NULL tests, but I don't think it can get any better. They verify that null is given as argument value to the query, and that's the closest I could get. After that, the most useful thing I could find was the parameterized query with ? and its bound arguments (the null value). I think after that point we can really only cross our fingers and presume the jdbc driver (or whatever) translates a null parameter to the null sql keyword internally, I don't think you can actually get the resulting sql string.

**Comment by stevenschlansker** [comment]
Don't worry about 100% unit test coverage.  As long as there's some sort of functional test making sure the feature behaves as specified that's likely good enough.

**Comment by TheRealMarnes** [comment]
Well yeah, but in this case we have a test that clearly shows a feature not working as intended. I'm honestly not sure what to do with that now. The tests show null is definitely bound to the query and debug view shows jdbc will execute a query with ? bound to null, yet the query doesn't seem to turn out as I expect. Unless h2 doesn't like "x in (null)" too much?

**Comment by stevenschlansker** [comment]
Oh, I misunderstood the problem.  It's entirely possible the behavior of that construct varies between h2 and pg.  I can take a bit of a look later tonight possibly if you haven't figure it out by then.

**Comment by TheRealMarnes** [comment]
Well, I did just boot up one of my own projects that runs an h2 db with mock data, and this happened:
vivaldi_2016-09-13_23-47-07

So it seems it's a problem with h2.

**Comment by stevenschlansker** [comment]
I think that looks correct?  `x IN (null)` should select no rows, even if some values are null, because `null != null`

**Comment by stevenschlansker** [comment]
http://stackoverflow.com/questions/6362112/in-clause-with-null-or-is-null

**Comment by TheRealMarnes** [comment]
Then I guess the tests actually mean it's working up to specs?...

I suppose what it really needs is an integration test with postgres instead of h2.

**Comment by TheRealMarnes** [comment]
Shall I just fix and break down the null handling test in light of this lesson, push it as such, and consider it finally done then? You guys can figure out a way to test the null handling some other time then, and in the meantime mark EmptyHandling.NULL as experimental/untested...

**Comment by stevenschlansker** [comment]
What exactly doesn't work right?  With `EmptyHandling.NULL`, if you pass in say a `new String[0]`, the behavior I expect is that it binds as `x IN (null)` and returns no rows.  Which part of this is broken?

File: src/test/java/org/skife/jdbi/v2/sqlobject/BindInNullTest.java [hunk 15]
```diff
@@ -0,0 +1,157 @@
...
+    }
+
+    @Test
+    public void testSomethingByIterableHandleNullWithEmptyList()
+    {
+        final SomethingByIterableHandleNull s = handle.attach(SomethingByIterableHandleNull.class);
+
+        final List<Something> out = s.get(new ArrayList<Object>());
+
+        Assert.assertEquals(4, out.size());
```
Comment (stevenschlansker) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInNullTest.java]: I expect `out` to be empty here, yes?

**Comment by TheRealMarnes** [comment]
Well I guess up to there it's working indeed. The thing I'd still like to verify is that it in fact does output the null keyword in the query so that it ends up being "x in (null)", but like I said earlier I don't think we can test for that...

Yeah sorry, my thoughts are a mess right now anyway. I guess everything is in order?

File: src/test/java/org/skife/jdbi/v2/sqlobject/BindInNullTest.java [hunk 16]
```diff
@@ -0,0 +1,157 @@
...
+        handle.execute("drop table something");
+        handle.close();
+    }
+
+    @Test
+    public void testSomethingByIterableHandleNullWithNull()
+    {
+        final SomethingByIterableHandleNull s = handle.attach(SomethingByIterableHandleNull.class);
+
+        final List<Something> out = s.get(null);
```
Comment (stevenschlansker) [inline src/test/java/org/skife/jdbi/v2/sqlobject/BindInNullTest.java]: This might be considered a programming error -- should we really assume `null` is `empty list` instead of a coding error?

**Comment by stevenschlansker** [comment]
Looking at the failing test, I think the test is wrong.  An empty input should match no rows no matter what data is in the table.

**Comment by TheRealMarnes** [comment]
> Looking at the failing test, I think the test is wrong. An empty input should match no rows no matter what data is in the table.
Yes, I'm changing that test atm. I mean the functionality itself is apparently working as intended.

**Comment by qualidafial** [comment]
> The thing I'd still like to verify is that it in fact does output the null keyword in the query so that it ends up being "x in (null)", but like I said earlier I don't think we can test for that...

You could `Mockito.spy` on the statement rewriter, and check that the rewritten query is as expected.

I feel this is getting out of hand though. I'm good to just merge this.

**Comment by stevenschlansker** [comment]
Yes, I think the existing test (once fixed) is sufficient.  No need for `spy`ing.

**Comment by TheRealMarnes** [comment]
agreed

**Comment by stevenschlansker** [comment]
Merged, thanks!

**Comment by TheRealMarnes** [comment]
Just for my information, around when are you planning to release 2.77? I kinda need the bindin fix in a project I'm working on (hence how I found the bug and why I did any of this to begin with)

**Comment by stevenschlansker** [comment]
@TheRealMarnes sorry was on vacation and lost track of it, just did the release :)

**Comment by TheRealMarnes** [comment]
Awesome, thanks
